src/daemon/plugin.c

```diff
@@ -1085,7 +1085,48 @@ static int plugin_insert_read(read_func_t *rf) {
   int status;
   llentry_t *le;
 
-  rf->rf_next_read = cdtime();
+  /* set the time of the first read according to "StartRead" in collectd.conf
+     start minute, second, millisecond is given as float in seconds */
+  double start = rf->rf_ctx.start_time; /* (default is -1.0) */
+  if (start < 0) {
+    /* if StartAt is not defined, start reading immediately */
+    rf->rf_next_read = cdtime();
+  } else {
+    /* get minute in hour and second+millisecond in minute */
+    cdtime_t now = cdtime();
+    struct timeval tv = CDTIME_T_TO_TIMEVAL(now);
+    uint64_t curr_msec =
+        (((tv.tv_sec / 60) % 60) * 60 // minute of hour in seconds
+         + gmtime(&(tv.tv_sec))->tm_sec) *
+            1000             // add seconds of minute and convert to ms
+        + tv.tv_usec / 1000; // add ms
+
+    /* convert to cdt time */
+    cdtime_t curr_cdt = MS_TO_CDTIME_T(curr_msec);
+    cdtime_t start_cdt = DOUBLE_TO_CDTIME_T(start);
+
+    /* NOTICE("%s: current time: %lu sec, %lu ms in minute, start time: %.3lf,
+    interval: %lu", rf->rf_name, tv.tv_sec, curr_msec, start,
+    CDTIME_T_TO_MS(rf->rf_interval)); */
+
+    /* get time to sleep until next read */
+    int64_t sleep_cdt = start_cdt - curr_cdt;
+    while (sleep_cdt < 0) {
+      sleep_cdt += rf->rf_interval;
+    }
+
+    /* set the next read time */
+    if (sleep_cdt > 0) {
+      rf->rf_next_read = now + sleep_cdt;
+      INFO("%s plugin: wait %.3fms before first read at %ld.%03ds since epoche",
```

Comment:
1) This will produce many similar lines when plugin registers several callbacks.
2) I don't think this message is useful for users, especially "123123123.123s since epoche" part.

IMHO: If you want, such message should be shown only once per plugin.

Reply:
I made it a debug message. I did not find an easy way to show it only once per plugin.